### PR TITLE
Keep pipe open while spout is running

### PR DESCRIPTION
**Pull request by adelelopez**
Fixes https://github.com/pachyderm/pachyderm/issues/4327

Named pipes have issues when the reading end is opened and closed rapidly (see https://stackoverflow.com/questions/47948952/named-pipe-dropping-data-in-c-program-using-fgets-in-linux). We realized that we didn't need to do it for spouts, so I simply moved the open and close out of the loop.

I also changed the kafka test to reproduce the bug.

### CODE REVIEW

File: etc/testing/kafka/main.go [hunk 1]
```diff
@@ -3,51 +3,82 @@ package main
...
+	// Set the default values of the configurable variables
+	var (
+		group_id = defaultGroupID
+		timeout  = defaultTimeout
+		pipe     = defaultNamedPipe
+	)
+
 	// And create a new kafka reader
+	fmt.Printf("creating new kafka reader for %v:%v with topic '%v' and group '%v'\n", host, port, topic, group_id)
+
```
Comment (jdoliner) [inline etc/testing/kafka/main.go]: Debug line that should probably be removed.

File: etc/testing/kafka/main.go [hunk 2]
```diff
@@ -3,51 +3,82 @@ package main
...
 	// this is the file loop
 	for {
 		if err := func() error {
-			tw := tar.NewWriter(out)
-			defer tw.Close()
-			// this is the message loop
 			// read a message
-			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-			defer cancel()
+			fmt.Printf("reading kafka queue.\n")
```
Comment (jdoliner) [inline etc/testing/kafka/main.go]: Debug line

File: etc/testing/kafka/main.go [hunk 3]
```diff
@@ -3,51 +3,70 @@ package main
...
-			tw := tar.NewWriter(out)
-			defer tw.Close()
-			// this is the message loop
 			// read a message
-			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-			defer cancel()
+			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
+			defer func() {
+				cancel()
+			}()
```
Comment (jdoliner) [inline etc/testing/kafka/main.go]:
```suggestion
			defer cancel()
```

Shorter version

File: etc/testing/kafka/main.go [hunk 4]
```diff
@@ -3,51 +3,70 @@ package main
...
+
+			// Open the /pfs/out pipe with write only permissons (the pachyderm spout will be reading at the other end of this)
+			// Note: it won't work if you try to open this with read, or read/write permissions
+			out, err := os.OpenFile(pipe, os.O_WRONLY, 0644)
+			if err != nil {
+				panic(err)
+			}
+			defer func() {
+				out.Close()
+			}()
```
Comment (jdoliner) [inline etc/testing/kafka/main.go]:
```suggestion
			defer out.Close()
```

File: etc/testing/kafka/main.go [hunk 5]
```diff
@@ -3,51 +3,70 @@ package main
...
+
+			tw := tar.NewWriter(out)
+			defer func() {
+				tw.Close()
+			}()
```
Comment (jdoliner) [inline etc/testing/kafka/main.go]:
```suggestion
			defer tw.Close()
```

File: src/server/worker/master.go [hunk 6]
```diff
@@ -255,6 +255,18 @@ func (a *APIServer) spoutSpawner(pachClient *client.APIClient) error {
 
 	var dir string
 
+	// Spouts typically have an open commit waiting for new data. So if the spout needs to be updated, and
+	// thus spoutSpawner is called, it might hang if the commit never gets closed. So to avoid this, we
+	// close open commits that we see here.
```
Comment (jdoliner) [inline src/server/worker/master.go:260]: Minor thing here, the comment is incorrect, we delete the commits below rather than closing them.

**Comment by adelelopez** [comment]
It only uses them in the kafka example code, so I don't think we want to use errors for that.

**Pull request merged**